Modules/_uuidmodule.c

```diff
@@ -91,45 +121,1730 @@ py_windows_has_stable_node(void)
+            (unsigned char *)self->bytes,
+            16,
+            0,  // big-endian
+            0,  // unsigned
+            1   // with_exceptions
+        ) < 0)
+    {
+        return -1;
+    }
+
+    return 0;
+}
+
+static int
+extract_field(
+    PyObject *fields,
+    int field_num,
+    uint64_t max_value,
+    const char *error_msg,
+    uint64_t *result
+) {
+    PyObject *field = PySequence_GetItem(fields, field_num);
+    if (field == NULL) {
+        return -1;
+    }
+
+    if (!PyLong_Check(field)) {
+        PyErr_Format(PyExc_TypeError, "field %d must be an integer", field_num);
+        goto fail;
+    }
+
+    uint64_t value;
+    if (PyLong_AsUInt64(field, &value) < 0) {
+        goto fail;
+    }
+
+    if (value > max_value) {
+        PyErr_SetString(PyExc_ValueError, error_msg);
+        goto fail;
+    }
+
+    *result = value;
+    Py_DECREF(field);
+    return 0;
+
+fail:
+    Py_DECREF(field);
+    return -1;
+}
+
+static int
+from_fields(uuidobject *self, PyObject *fields)
+{
+    // Validate that fields is a sequence with exactly 6 elements
+    if (!PySequence_Check(fields)) {
+        PyErr_SetString(PyExc_TypeError, "fields must be a sequence");
+        return -1;
+    }
+
+    Py_ssize_t len = PySequence_Size(fields);
+    if (len != 6) {
+        PyErr_SetString(PyExc_ValueError, "fields is not a 6-tuple");
+        return -1;
+    }
+
+#   define EXTRACT_FIELD(field_num, max_value, error_msg, type, name)           \
+        type name;                                                              \
+        do {                                                                    \
+            uint64_t name##_extracted;                                          \
+            if (extract_field(fields, field_num, max_value, error_msg,          \
+                            &(name##_extracted)) < 0) {                         \
+                return -1;                                                      \
+            }                                                                   \
+            name = (type)name##_extracted;                                      \
+        } while(0)
+
+    EXTRACT_FIELD(
+        0, (1ULL << 32) - 1, "field 1 out of range (need a 32-bit value)",
+        uint32_t, time_low
+    );
+    EXTRACT_FIELD(
+        1, (1ULL << 16) - 1, "field 2 out of range (need a 16-bit value)",
+        uint16_t, time_mid
+    );
+    EXTRACT_FIELD(
+        2, (1ULL << 16) - 1, "field 3 out of range (need a 16-bit value)",
+        uint16_t, time_hi_version
+    );
+    EXTRACT_FIELD(
+        3, (1ULL << 8) - 1, "field 4 out of range (need an 8-bit value)",
+        uint8_t, clock_seq_hi_variant
+    );
+    EXTRACT_FIELD(
+        4, (1ULL << 8) - 1, "field 5 out of range (need an 8-bit value)",
+        uint8_t, clock_seq_low
+    );
+    EXTRACT_FIELD(
+        5, (1ULL << 48) - 1, "field 6 out of range (need a 48-bit value)",
+        uint64_t, node
+    );
+
```

Comment:
```suggestion
        uint64_t, node
    );

#undef EXTRACT_FIELD
```

I tend to undef local macros (as per PEP-7 guidance) to reduce the possibility of using it outside its scope.